### PR TITLE
#488 added onInitComplete for CLI application module initialization

### PR DESCRIPTION
**Pull request by belisoful**
This is an addition-enhancement that will help CLI applications initialize after the modules are loaded up.  It replaces OnBeginRequest as the first application event.  Shell applications do not have an application event stack like normal applications... but they still should have post module initialization.

Use Case, A Module that uses the the database module may want to do its database work in onInitComplete if, in case, the database module is loaded after the A Module.

### CODE REVIEW

File: framework/TApplication.php [hunk 1]
```diff
@@ -1023,6 +1024,7 @@ public function applyConfiguration($config, $withinService = false)
 	 * Loads configuration and initializes application.
 	 * Configuration file will be read and parsed (if a valid cached version exists,
 	 * it will be used instead). Then, modules are created and initialized;
+	 * Finalize any modules by calling onInitComplete, mainly for CLI initialization.
```
Comment (ctrlaltca) [inline framework/TApplication.php:1027]: Please reword this; onInitComplete is an event, so the use doesn't "call" it.
Something like "module initialization can be finalized in the onInitComplete event, this is particularly useful for CLI applications"

File: framework/TApplication.php [hunk 2]
```diff
@@ -1104,6 +1108,20 @@ public function onError($param)
 		$this->getErrorHandler()->handleError($this, $param);
 	}
 
+	/**
+	 * Raises onInitComplete event.
+	 * At the time when this method is invoked, application modules are loaded,
+	 * user request is resolved and the corresponding service is loaded and
+	 * initialized. The application is about to start processing the user
```
Comment (ctrlaltca) [inline framework/TApplication.php:1115]: At this step services are loaded  but not initialized yet: they have no ID assigned and their still have to load their config.

File: framework/TApplication.php [hunk 3]
```diff
@@ -1104,6 +1108,20 @@ public function onError($param)
...
+	 * request.  This call is important for Shell applications that do not have
+	 * a web service stack.  This is the first and last call for finalization
+	 * of any loaded modules in Shell mode.
```
Comment (ctrlaltca) [inline framework/TApplication.php:1118]: Sorry for being this nitpick, but I find this a bit confusing for new users,  can you please reword this to make it more propositive, something like "In this event modules can perform any needed late initialization before the service starts to process the request" or something similar?

**Review by ctrlaltca — CHANGES_REQUESTED** [review]
Hi, and first thank you for your PR.
The change itself looks fine, please have a look at the few possible improvements on the comments.

**Comment by belisoful** [comment]
Yeah.  I'm not exactly great at the commenting thing yet.  Getting there.  You aren't Nitpicking...  Your comments are exactly what the code comments need!  I'm looking for feedback, so THANK YOU SO MUCH!!!   I'l make the changes and get back to you shortly.

**Comment by belisoful** [comment]
My humble apologies, I'm beginning to think that onInitComplete should be after the service initialization.  Your thoughts?

**Comment by belisoful** [comment]
BTW, I cannot describe how much I (and we) appreciate your upkeep of PRADO all this time.

**Comment by ctrlaltca** [comment]
> My humble apologies, I'm beginning to think that onInitComplete should be after the service initialization. Your thoughts?

Maybe we could just make `TApplication::_steps` protected and override it in `TShellApplication::_steps` providing a few common steps that can be usable on a CLI application. Then `TShellApplication::run()` will just loop on the events like `TApplication::run()` already does.
What do you think of this approach?

**Comment by belisoful** [comment]
From the standpoint of shell user input, looping on (a reduced number of) steps may be useful.  if the CLI is processing single commands, looping wouldn't be needed.  The CLI doesn't exactly have requests in the same way that the web app/service does.  unifying these steps between these differing app types would be of great help to modules. but does it make sense?  I'll have to look into the Shell App and what it's functionality is actually doing.

**Comment by belisoful** [comment]
I have a whole slew of new functionality for Prado in mind.  I am writing tickets to address the short comings, their fixes, and hopefully quickly, checkins.

**Comment by belisoful** [comment]
BTW, ctrlaltca, I am a member of the PRADO group under the hidden name javalizard.   While I could log in under my old user name and do the code changes myself, I thought my own branch may be better given my soon to be proposed changes/additions for PRADO. 

There were a number of serious flaws I found in my architectural use of PRADO and that is the source of these recent issues and changes.

The hope is to have these changes included in the 4.2,0 release.  Soon, I will look at #339 and make my changes to the system for enabling auto global listen.  It will be generally turned on, and some subclasses may have them turned off, like Lists, Maps, etc, collections.  It will be a general review of the __construct method esp. as relates to auto global listen.

**Comment by belisoful** [comment]
It has taken until PHP 7.4 to really get Class Behaviors technically proper with WeakReferences.  A memory leak that big could be a problem for systems, which is why it has been off.   but with PHP 7.4, auto global listen can be turned on without the memory leaks from the lack of weakReferences.

File: framework/Web/Services/TPageService copy.php [hunk 4]
```diff
@@ -0,0 +1,573 @@
+<?php
```
Comment (ctrlaltca) [inline framework/Web/Services/TPageService copy.php:1]: This file has probably been pushed by mistake

File: framework/Web/UI/WebControls/TPageLoadTime.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * TPageLoadTime class.
+ *
+ * Writes the amount of time taken from application start to rendering the contents of this control.
+ *
+ * @author Brad Anderson <belisoful@icloud.com>
+ * @package Prado\Web\UI\WebControls
+ * @since 4.2.0
+ */
+class TPageLoadTime extends TLabel
```
Comment (ctrlaltca) [inline framework/Web/UI/WebControls/TPageLoadTime.php:22]: Do we really need a web control for this? TLogger already does the same, and its output can be routed to different places as needed

**Comment by ctrlaltca** [comment]
Please, if you are implementing different changes, use different branches and create multiple PRs.
Eg. the TPageLoadTime changes are unrelated to the rest, they can be made in a different branch

File: framework/Web/Services/TPageService copy.php (hunk 4 above)
Comment (belisoful) [reply framework/Web/Services/TPageService copy.php:1]: Indeed I was!  I don't know how this got in there.  I specifically did not check this file

File: framework/Web/UI/WebControls/TPageLoadTime.php (hunk 5 above)
Comment (belisoful) [reply framework/Web/UI/WebControls/TPageLoadTime.php:22]: There are webpages that link in the load time.  It was to make use of the application start time.

**Comment by belisoful** [comment]
If you believe the startTime and initial memory usage are not good, I'll reverse that for you on this branch.  Some websites like to post the page load time at the bottom (or somewhere) of the page.  I don't know why they do that, but the TPageLoadTime is a use case for the application capturing its start time.

If I remember correctly, most platforms record their start time.

**Comment by belisoful** [comment]
I am working on unifying the __construct methods' parent chain throughout the framework.  I am hoping this __construct change gets into the 4.2.0 release.  I am verifying my work today and running tests after.

**Comment by belisoful** [comment]
I don't know what GitHub was doing. I only had one file selected for checkin.  It started checking in all the parent::__construct changes too!  I didn't mean for it to do that.  That is supposed to be a single large checkin.  I feel bad now.  making myself look back not using these tools right.  sigh.  ok.  back to work.

**Comment by belisoful** [comment]
This set of updates is complete.  Have a look across the changes for the pull request, if you don't mind.

**Comment by belisoful** [comment]
I'm working on getting these checks to complete.  All of these changes should be pulled.  They make the framework much more competitive and adaptable.  The CLI Application commands are important.  I'm working on a few other things.

**Comment by belisoful** [comment]
not sure what the braces or binary_operator_spaces errors are coming from.

**Comment by ctrlaltca** [comment]
This is becoming a really big change :) I'm not sure how to deal with it.
I'd like to merge at least the __construct changes soon, but i don't want to ruin your branch by cherry picking.
Do you mind if I ask you to move all the __construct changes in another new PR so that i can review, test and merge them?
We can then proceed to review and merge all the other changes the same way, one functional piece at a time.

**Comment by belisoful** [comment]
these are changes I've been hanging onto for a while and should be folded back.  I've been testing the crap out of these things.  There are more changes forth-coming too, and also, hopefully, more discussion.  This new functionality should be a winner for the 4.2 release.  The changed file list is large due to the __construct update.  Dealing with that seems like a good first step.

It is a lot of functionality very quickly.  Sorry about that.  It's just, I've been sitting on this gold mine for a while and it should be shared.

I'll make a new PR with select changes.  I'll get back with you shortly with the new PR.

I have an account as a PRADO developer under the name Javalizard.  I just don't use that any more. Would it be more useful to be doing the merging myself in a controlled way?  like transfer over the changes by selected commits based on functionality?  If you need proof that I am Javalizard, I can log in to GitHub under that name and set a few of the issues to bugs.  We have some options.  just saying.

**Comment by belisoful** [comment]
We'll talk on the other pull.  This pull is closed.
